File: DynamicKey/AgoraDynamicKey/php/test/ChatTokenBuilder2Test.php
<?php

require_once "../src/AccessToken2.php";
require_once "../src/ChatTokenBuilder2.php";

class ChatTokenBuilder2Test
{
    public $appId = "970CA35de60c44645bbae8a215061b33";
    public $appCertificate = "5CFd2fd1755d40ecb72977518be15d3b";
    public $expire = 600;
    public $userId = "2882341273";

    public function run()
    {
        $this->test_buildUserToken();
        $this->test_buildAppToken();
    }

    public function test_buildUserToken()
    {
        $token = ChatTokenBuilder2::buildUserToken($this->appId, $this->appCertificate, $this->userId, $this->expire);
        $accessToken = new AccessToken2();
        $accessToken->parse($token);

        Util::assertEqual($this->appId, $accessToken->appId);
        Util::assertEqual($this->expire, $accessToken->expire);
        Util::assertEqual(ServiceChat::SERVICE_TYPE, $accessToken->services[ServiceChat::SERVICE_TYPE]->type);
        Util::assertEqual($this->userId, $accessToken->services[ServiceChat::SERVICE_TYPE]->userId);
        Util::assertEqual($this->expire, $accessToken->services[ServiceChat::SERVICE_TYPE]->privileges[ServiceChat::PRIVILEGE_USER]);
    }

    public function test_buildAppToken()
    {
        $token = ChatTokenBuilder2::buildAppToken($this->appId, $this->appCertificate, $this->expire);
        $accessToken = new AccessToken2();
        $accessToken->parse($token);

        Util::assertEqual($this->appId, $accessToken->appId);
        Util::assertEqual($this->expire, $accessToken->expire);
        Util::assertEqual(ServiceChat::SERVICE_TYPE, $accessToken->services[ServiceChat::SERVICE_TYPE]->type);
        Util::assertEqual($this->expire, $accessToken->services[ServiceChat::SERVICE_TYPE]->privileges[ServiceChat::PRIVILEGE_APP]);
    }
}

$chatTokenBuilder2Test = new ChatTokenBuilder2Test();
$chatTokenBuilder2Test->run();

File: DynamicKey/AgoraDynamicKey/cpp/test/ChatTokenBuilder2_test.cpp
// Copyright (c) 2014-2017 Agora.io, Inc.
//

#define private public
#define protected public

#include "../src/ChatTokenBuilder2.h"

#include <gtest/gtest.h>

#include "../src/AccessToken2.h"

using namespace agora::tools;

class ChatTokenBuilder2_test : public testing::Test {
 protected:
  virtual void SetUp() override {
    app_id_ = "970CA35de60c44645bbae8a215061b33";
    app_cert_ = "5CFd2fd1755d40ecb72977518be15d3b";
    user_id_ = "test_user";
    expire_ = 600;
  }

  void TestChatTokenBuilderBuildUserToken() {
    std::string token = ChatTokenBuilder2::BuildUserToken(app_id_,
      app_cert_, user_id_, expire_);

    AccessToken2 parser;
    bool parsed = parser.FromString(token);

    EXPECT_EQ(parser.signature_, parser.GenerateSignature(app_cert_));

    ASSERT_TRUE(parsed);
    EXPECT_EQ(parser.app_id_, app_id_);
    EXPECT_EQ(parser.expire_, expire_);
    EXPECT_EQ(parser.services_.size(), 1);

    Service *service = parser.services_.begin()->second.get();

    EXPECT_EQ(dynamic_cast<ServiceChat *>(service)->user_id_, user_id_);
    EXPECT_EQ(service->privileges_[ServiceChat::kPrivilegeUser], expire_);
  }


  void TestChatTokenBuilderBuildAppToken() {
    std::string token = ChatTokenBuilder2::BuildAppToken(app_id_,
      app_cert_, expire_);

    AccessToken2 parser;
    bool parsed = parser.FromString(token);

    EXPECT_EQ(parser.signature_, parser.GenerateSignature(app_cert_));

    ASSERT_TRUE(parsed);
    EXPECT_EQ(parser.app_id_, app_id_);
    EXPECT_EQ(parser.expire_, expire_);
    EXPECT_EQ(parser.services_.size(), 1);

    Service *service = parser.services_.begin()->second.get();

    EXPECT_EQ(service->privileges_[ServiceChat::kPrivilegeApp], expire_);
  }

 private:
  std::string app_id_;
  std::string app_cert_;
  std::string user_id_;
  uint32_t expire_;
};

TEST_F(ChatTokenBuilder2_test, testChatTokenBuilderBuildUserToken) { TestChatTokenBuilderBuildUserToken(); }
TEST_F(ChatTokenBuilder2_test, testChatTokenBuilderBuildAppToken) { TestChatTokenBuilderBuildAppToken(); }
